prov/util/src/util_cq.c

```diff
@@ -301,11 +301,20 @@ ssize_t ofi_cq_sread(struct fid_cq *cq_fid, void *buf, size_t count,
 		const void *cond, int timeout)
 {
 	struct util_cq *cq;
+	int ret;
 
 	cq = container_of(cq_fid, struct util_cq, cq_fid);
 	assert(cq->wait && cq->internal_wait);
-	fi_wait(&cq->wait->wait_fid, timeout);
-	return ofi_cq_read(cq_fid, buf, count);
+
+	while (1) {
+		ret = fi_wait(&cq->wait->wait_fid, timeout);
```

Comment:
We need to reset the timeout, or we'll end up asking to wait for the full wait timeout again if we break out early.  The adjustments to the timeout done within fi_wait are lost on exit from that call.

nit: You could restructure the outer loop as do { } while (ret == -FI_EAGAIN)

Reply:
Hmm. Seems like we can as well add the timeout check here.